vpr/src/noc/sat_routing.cpp

```diff
@@ -0,0 +1,986 @@
+
+    auto router_it = std::find_if(noc_routers.begin(), noc_routers.end(), [noc_router_latency](const NocRouter& r) {
+        return (noc_router_latency != r.get_latency());
+    });
+
+    if (router_it != noc_routers.end()) {
+        VTR_LOG_ERROR(
+            "SAT router assumes all NoC routers have the same latency. "
+            "NoC router with the user if %d has a different latency (%g) than the NoC-wide router latency (%g).\n",
+            router_it->get_router_user_id(), router_it->get_latency());
+    }
+
+    auto link_it = std::find_if(noc_links.begin(), noc_links.end(), [noc_link_latency](const NocLink& l) {
+       return (noc_link_latency != l.get_latency());
+    });
+
+    if (link_it != noc_links.end()) {
+        VTR_LOG_ERROR(
+            "SAT router assumes all NoC links have the same latency. "
+            "NoC link %d has a different latency (%g) than the NoC-wide link latency (%g).\n",
+            (size_t)link_it->get_link_id(), link_it->get_latency());
+    }
+
+    const double traffic_flow_latency_constraint = traffic_flow.max_traffic_flow_latency;
+
+    VTR_ASSERT(traffic_flow_latency_constraint < 0.1);
+
+    int n_max_links = std::floor((traffic_flow_latency_constraint - noc_router_latency) / (noc_link_latency + noc_router_latency));
+
+    return n_max_links;
+}
+
+static void constrain_latency_overrun_vars(orsat::CpModelBuilder& cp_model,
+                                           t_flow_link_var_map& flow_link_vars,
+                                           std::map<NocTrafficFlowId , orsat::IntVar>& latency_overrun_vars) {
+    const auto& noc_ctx = g_vpr_ctx.noc();
+    const auto& noc_model = noc_ctx.noc_model;
+
+    for (auto& [traffic_flow_id, latency_overrun_var] : latency_overrun_vars) {
+        int n_max_links = comp_max_number_of_traversed_links(traffic_flow_id);
+        // get all boolean variables for this traffic flow
+        auto link_vars = get_flow_link_vars(flow_link_vars, {traffic_flow_id},
+                                            {noc_model.get_noc_links().keys().begin(), noc_model.get_noc_links().keys().end()});
+
+        orsat::LinearExpr latency_overrun_expr;
+        // count the number of activated links for this traffic flow
+        latency_overrun_expr += orsat::LinearExpr::Sum(link_vars);
+        // subtract the maximum number of permissible links from the number of activated links
+        latency_overrun_expr -= n_max_links;
+
+        // if latency_overrun_expr is non-positive, the latency constraint is met and
+        // latency_overrun_var should be zero. Otherwise, should be equal to latency_overrun_expr
+        // This is like pushing latency_overrun_expr through a ReLU function to get latency_overrun_var.
+        cp_model.AddMaxEquality(latency_overrun_var, {latency_overrun_expr, 0});
+    }
+}
+
+//static void add_movable_continuity_constraints(t_flow_link_var_map& flow_link_vars,
```

Comment:
Comment that this is for SAT NoC placement; move to end of file in a big block with other routines for that.